Design/spectrum_number_and_workspace_index_abstraction.md

```diff
@@ -0,0 +1,269 @@
+# Indexing Abstraction Layer
```

Comment:
If we haven't already. Maybe it is worth stating some of the guarantees about what `index` means. integer set starting from 0 etc. @SimonHeybrock I prefer the new terminology. Yes "workspace index" can be a confusing one, particularly when working with groups.